src/nat/libp2p_nat.erl

```diff
@@ -0,0 +1,212 @@
+%%%-------------------------------------------------------------------
+%% @doc
+%% == Libp2p NAT ==
+%% @end
+%%%-------------------------------------------------------------------
+-module(libp2p_nat).
+
+%% ------------------------------------------------------------------
+%% API Function Exports
+%% ------------------------------------------------------------------
+-export([
+    enabled/1
+    ,maybe_spawn_discovery/3, spawn_discovery/3
+    ,add_port_mapping/2
+    ,maybe_apply_nat_map/1
+]).
+
+-ifdef(TEST).
+-include_lib("eunit/include/eunit.hrl").
+-endif.
+
+%%--------------------------------------------------------------------
+%% @doc
+%% @end
+%%--------------------------------------------------------------------
+-spec enabled(ets:tab() | list()) -> boolean().
+enabled(Opts) when is_list(Opts) ->
+    libp2p_config:get_opt(Opts, [?MODULE, nat], true);
+enabled(TID) ->
+    Opts = libp2p_swarm:opts(TID),
+    libp2p_config:get_opt(Opts, [?MODULE, nat], true).
+
+%%--------------------------------------------------------------------
+%% @doc
+%% @end
+%%--------------------------------------------------------------------
+-spec maybe_spawn_discovery(pid(), [string()], ets:tab()) -> ok.
+maybe_spawn_discovery(Pid, MultiAddrs, TID) ->
+    case ?MODULE:enabled(TID) of
+        true ->
+            erlang:spawn(?MODULE, spawn_discovery, [Pid, MultiAddrs, TID]),
+            ok;
+        false ->
+            lager:warning("nat is disabled")
+    end.
+
+%%--------------------------------------------------------------------
+%% @doc
+%% @end
+%%--------------------------------------------------------------------
+-spec spawn_discovery(pid(), [string()], ets:tab()) -> ok.
+spawn_discovery(Pid, MultiAddrs, _TID) ->
+    case lists:filtermap(fun discovery_filter/1, MultiAddrs) of
+        [] -> ok;
+        [{MultiAddr, _IP, Port}|_] ->
+            lager:info("trying to map ~p ~p", [MultiAddr, Port]),
+            %% TODO we should make a port mapping for EACH address
```

Comment:
This could be added btw, let me know

Reply:
Yeah, that would definitely help us as we expect to have our gateway hardware have several network interfaces active simultaneously.